src/pulsar_producer.erl

```diff
@@ -49,233 +94,687 @@ callback_mode() -> [state_functions].
+         end,
+         {SequenceId, 0},
+         TSMsgs)
+      end,
+      Requests);
+code_change_requests(_Vsn, Requests) ->
+    %% The upgrade back from the old version is lossy; we can't
+    %% produce an `replayq:ack_ref', the messages themselves, or the
+    %% `Froms'...  Have to handle this case when resending or
+    %% receiving a response.
+    Requests.
+
+%% -record(state, {partitiontopic,
+%%                 broker_server,
+%%                 sock,
+%%                 request_id = 1,
+%%                 producer_id = 1,
+%%                 sequence_id = 1,
+%%                 producer_name,
+%%                 opts = #{},
+%%                 callback,
+%%                 batch_size = 0,
+%%                 requests = #{},
+%%                 last_bin = <<>>}).
+to_old_state_record(StateMap = #{}) ->
+    { state
+    , maps:get(partitiontopic, StateMap)
+    , maps:get(broker_server, StateMap)
+    , maps:get(sock, StateMap)
+    , maps:get(request_id, StateMap)
+    , maps:get(producer_id, StateMap)
+    , maps:get(sequence_id, StateMap)
+    , maps:get(producer_name, StateMap)
+    , maps:get(opts, StateMap)
+    , maps:get(callback, StateMap)
+    , maps:get(batch_size, StateMap)
+    , maps:get(requests, StateMap)
+    , maps:get(last_bin, StateMap)
+    }.
+
+from_old_state_record(StateRec) ->
+    #{ partitiontopic => element(2, StateRec)
+     , broker_server => element(3, StateRec)
+     , sock => element(4, StateRec)
+     , request_id => element(5, StateRec)
+     , producer_id => element(6, StateRec)
+     , sequence_id => element(7, StateRec)
+     , producer_name => element(8, StateRec)
+     , opts => element(9, StateRec)
+     , callback => element(10, StateRec)
+     , batch_size => element(11, StateRec)
+     , requests => element(12, StateRec)
+     , last_bin => element(13, StateRec)
+     }.
+
+-spec escape(string()) -> binary().
+escape(Str) ->
+    NormalizedStr = unicode:characters_to_nfd_list(Str),
+    iolist_to_binary(escape_uri(NormalizedStr)).
+
+%% copied from `edoc_lib' because dialyzer cannot see this private
+%% function there.
+escape_uri([C | Cs]) when C >= $a, C =< $z ->
+    [C | escape_uri(Cs)];
+escape_uri([C | Cs]) when C >= $A, C =< $Z ->
+    [C | escape_uri(Cs)];
+escape_uri([C | Cs]) when C >= $0, C =< $9 ->
+    [C | escape_uri(Cs)];
+escape_uri([C = $. | Cs]) ->
+    [C | escape_uri(Cs)];
+escape_uri([C = $- | Cs]) ->
+    [C | escape_uri(Cs)];
+escape_uri([C = $_ | Cs]) ->
+    [C | escape_uri(Cs)];
+escape_uri([C | Cs]) when C > 16#7f ->
+    %% This assumes that characters are at most 16 bits wide.
+    escape_byte(((C band 16#c0) bsr 6) + 16#c0)
+	++ escape_byte(C band 16#3f + 16#80)
+	++ escape_uri(Cs);
+escape_uri([C | Cs]) ->
+    escape_byte(C) ++ escape_uri(Cs);
+escape_uri([]) ->
+    [].
+
+%% copied from `edoc_lib' because dialyzer cannot see this private
+%% function there.
+escape_byte(C) when C >= 0, C =< 255 ->
+    [$%, hex_digit(C bsr 4), hex_digit(C band 15)].
```

Comment:
Windows system does not allow `%` in file path, use `=` instead. 
```suggestion
    [$=, hex_digit(C bsr 4), hex_digit(C band 15)].
```

Reply:
or `_` should do too.

Reply:
Oops, I copied the fn and forgot that final detail :see_no_evil:

Reply:
:heavy_check_mark: 

Used a `_`